File: server/src/mcp/services/NotionService.ts
import { BaseMCPService } from '../BaseMCPService.js';
import type {
  MCPServiceConfig,
  MCPOperationRequest,
  MCPOperationResult,
} from '../types.js';

/**
 * Notion MCP 서비스
 *
 * 허용 작업:
 * - 페이지 읽기/검색
 * - 페이지 초안 생성 (승인 필요)
 * - 페이지 업데이트 (승인 필요)
 *
 * 금지 작업:
 * - 승인 없는 페이지 공개
 * - 승인 없는 공유 설정 변경
 */
export class NotionService extends BaseMCPService {
  private apiKey?: string;

  constructor(config: MCPServiceConfig) {
    super({
      ...config,
      type: 'notion',
    });
    this.apiKey = config.credentials?.apiKey;
  }

  protected async doConnect(): Promise<void> {
    if (!this.apiKey) {
      throw new Error('Notion API key is required');
    }

    // TODO: 실제 Notion API 연결 구현
    console.log(`[NotionService] Connected to Notion`);
  }

  protected async doDisconnect(): Promise<void> {
    console.log(`[NotionService] Disconnected from Notion`);
  }

  protected async doExecute(request: MCPOperationRequest): Promise<MCPOperationResult> {
    const { operation, target, payload } = request;

    switch (operation) {
      case 'read':
        return this.readPage(target.id!);

      case 'search':
        return this.searchPages(payload as { query: string });

      case 'list':
        return this.listPages(target.path);

      case 'create':
        return this.createPage(target.path!, payload as Record<string, unknown>);

      case 'update':
        return this.updatePage(target.id!, payload as Record<string, unknown>);

      case 'delete':
        return this.archivePage(target.id!);

      default:
        return {
          success: false,
          error: `Unsupported operation: ${operation}`,
        };
    }
  }

  protected async doValidate(request: MCPOperationRequest): Promise<{
    errors: string[];
    warnings: string[];
  }> {
    const errors: string[] = [];
    const warnings: string[] = [];

    // 작업별 검증
    switch (request.operation) {
      case 'read':
      case 'update':
      case 'delete':
        if (!request.target?.id) {
          errors.push('Page ID is required');
        }
        break;

      case 'create':
        if (!request.target?.path) {
          errors.push('Parent page path is required');
        }
        if (!request.payload?.title) {
          warnings.push('Page title is recommended');
        }
        break;

      case 'search':
        if (!(request.payload as { query?: string })?.query) {
          errors.push('Search query is required');
        }
        break;
    }

    return { errors, warnings };
  }

  protected shouldRequireApproval(request: MCPOperationRequest): boolean {
    // Notion에서는 create, update도 승인 필요
    return ['create', 'update'].includes(request.operation);
  }

  // === Notion API 작업 구현 ===

  private async readPage(pageId: string): Promise<MCPOperationResult> {
    // TODO: 실제 Notion API 호출
    console.log(`[NotionService] Reading page: ${pageId}`);
    return {
      success: true,
      data: {
        id: pageId,
        title: 'Sample Page',
        content: '페이지 내용...',
      },
    };
  }

  private async searchPages(params: { query: string }): Promise<MCPOperationResult> {
    console.log(`[NotionService] Searching pages: ${params.query}`);
    return {
      success: true,
      data: {
        results: [],
        hasMore: false,
      },
    };
  }

  private async listPages(parentPath?: string): Promise<MCPOperationResult> {
    console.log(`[NotionService] Listing pages in: ${parentPath || 'root'}`);
    return {
      success: true,
      data: {
        pages: [],
      },
    };
  }

  private async createPage(
    parentPath: string,
    payload: Record<string, unknown>
  ): Promise<MCPOperationResult> {
    console.log(`[NotionService] Creating page in: ${parentPath}`);
    return {
      success: true,
      data: {
        id: 'new-page-id',
        url: 'https://notion.so/new-page',
        ...payload,
      },
    };
  }

  private async updatePage(
    pageId: string,
    payload: Record<string, unknown>
  ): Promise<MCPOperationResult> {
    console.log(`[NotionService] Updating page: ${pageId}`);
    return {
      success: true,
      data: {
        id: pageId,
        ...payload,
      },
    };
  }

  private async archivePage(pageId: string): Promise<MCPOperationResult> {
    console.log(`[NotionService] Archiving page: ${pageId}`);
    return {
      success: true,
      data: {
        id: pageId,
        archived: true,
      },
    };
  }
}

File: server_python/mcp/services/notion_service.py
from typing import Dict, Any, Optional
from ..base_mcp_service import BaseMCPService
from ..types import (
    MCPServiceConfig,
    MCPOperationRequest,
    MCPOperationResult,
    MCPOperationType,
)


class NotionService(BaseMCPService):
    """
    Notion MCP 서비스
    
    허용 작업:
    - 페이지 읽기/검색
    - 페이지 초안 생성 (승인 필요)
    - 페이지 업데이트 (승인 필요)
    
    금지 작업:
    - 승인 없는 페이지 공개
    - 승인 없는 공유 설정 변경
    """
    
    def __init__(self, config: MCPServiceConfig):
        super().__init__(config)
        self.api_key = config.credentials.get("apiKey") if config.credentials else None
    
    async def _do_connect(self) -> None:
        if not self.api_key:
            raise ValueError("Notion API key is required")
        # TODO: 실제 Notion API 연결 구현
        print(f"[NotionService] Connected to Notion")
    
    async def _do_disconnect(self) -> None:
        print(f"[NotionService] Disconnected from Notion")
    
    async def _do_execute(self, request: MCPOperationRequest) -> MCPOperationResult:
        operation = request.operation
        target = request.target
        payload = request.payload
        
        if operation == MCPOperationType.READ:
            return await self._read_page(target.id)
        elif operation == MCPOperationType.SEARCH:
            return await self._search_pages(payload)
        elif operation == MCPOperationType.LIST:
            return await self._list_pages(target.path)
        elif operation == MCPOperationType.CREATE:
            return await self._create_page(target.path, payload)
        elif operation == MCPOperationType.UPDATE:
            return await self._update_page(target.id, payload)
        elif operation == MCPOperationType.DELETE:
            return await self._archive_page(target.id)
        else:
            return MCPOperationResult(
                success=False,
                error=f"Unsupported operation: {operation}"
            )
    
    async def _do_validate(self, request: MCPOperationRequest) -> Dict[str, list]:
        errors = []
        warnings = []
        
        operation = request.operation
        
        if operation in [MCPOperationType.READ, MCPOperationType.UPDATE, MCPOperationType.DELETE]:
            if not request.target.id:
                errors.append("Page ID is required")
        
        elif operation == MCPOperationType.CREATE:
            if not request.target.path:
                errors.append("Parent page path is required")
            if not request.payload.get("title"):
                warnings.append("Page title is recommended")
        
        elif operation == MCPOperationType.SEARCH:
            if not request.payload.get("query"):
                errors.append("Search query is required")
        
        return {"errors": errors, "warnings": warnings}
    
    def _should_require_approval(self, request: MCPOperationRequest) -> bool:
        # Notion에서는 create, update도 승인 필요
        return request.operation in [MCPOperationType.CREATE, MCPOperationType.UPDATE]
    
    # === Notion API 작업 구현 ===
    
    async def _read_page(self, page_id: str) -> MCPOperationResult:
        # TODO: 실제 Notion API 호출
        print(f"[NotionService] Reading page: {page_id}")
        return MCPOperationResult(
            success=True,
            data={
                "id": page_id,
                "title": "Sample Page",
                "content": "페이지 내용...",
            }
        )
    
    async def _search_pages(self, params: Dict[str, Any]) -> MCPOperationResult:
        print(f"[NotionService] Searching pages: {params.get('query')}")
        return MCPOperationResult(
            success=True,
            data={
                "results": [],
                "hasMore": False,
            }
        )
    
    async def _list_pages(self, parent_path: Optional[str] = None) -> MCPOperationResult:
        print(f"[NotionService] Listing pages in: {parent_path or 'root'}")
        return MCPOperationResult(
            success=True,
            data={"pages": []}
        )
    
    async def _create_page(self, parent_path: str, payload: Dict[str, Any]) -> MCPOperationResult:
        print(f"[NotionService] Creating page in: {parent_path}")
        return MCPOperationResult(
            success=True,
            data={
                "id": "new-page-id",
                "url": "https://notion.so/new-page",
                **payload,
            }
        )
    
    async def _update_page(self, page_id: str, payload: Dict[str, Any]) -> MCPOperationResult:
        print(f"[NotionService] Updating page: {page_id}")
        return MCPOperationResult(
            success=True,
            data={
                "id": page_id,
                **payload,
            }
        )
    
    async def _archive_page(self, page_id: str) -> MCPOperationResult:
        print(f"[NotionService] Archiving page: {page_id}")
        return MCPOperationResult(
            success=True,
            data={
                "id": page_id,
                "archived": True,
            }
        )

